cmd/kubeadm/app/cmd/upgrade/apply.go

```diff
@@ -0,0 +1,192 @@
+	cmd.Flags().BoolVar(&flags.dryRun, "dry-run", flags.dryRun, "Do not change any state, just output what actions would be applied. This also implies non-interactive mode")
+	cmd.Flags().StringVar(&flags.newK8sVersionStr, "version", flags.newK8sVersionStr, "Selected version to upgrade to")
+	cmd.Flags().DurationVar(&flags.pullTimeout, "pull-timeout", flags.pullTimeout, "The maximum amount of time to wait for the control plane pods to be downloaded")
+
+	return cmd
+}
+
+// RunApply takes care of the actual upgrade functionality
+func RunApply(flags *applyFlags) error {
+
+	// Start with the basics, verify that the cluster is healthy and get the configuration from the cluster (using the ConfigMap)
+	client, cfg, err := EnforceRequirements(flags.parent.kubeConfigPath, flags.parent.cfgPath, flags.parent.printConfig, flags.nonInteractiveMode)
+	if err != nil {
+		return err
+	}
+
+	// Set the upgraded version on the config object now
+	cfg.KubernetesVersion = flags.newK8sVersionStr
+
+	// Grab the external, versioned configuration and convert it to the internal type for usage here later
+	internalcfg := &kubeadmapi.MasterConfiguration{}
+	api.Scheme.Convert(cfg, internalcfg, nil)
+
+	// Enforce all the version skew policies there are
+	if err := EnforceVersionPolicies(flags, client); err != nil {
+		return err
+	}
+
+	// If interactive mode is set (--yes was not specified), ask ther user whether they really want to upgrade
+	// If --yes is specified; just proceed and do it
+	if !flags.nonInteractiveMode {
+		if err := upgrade.InteractivelyConfirmUpgrade("Are you sure you want to proceed with the upgrade"); err != nil {
+			return err
+		}
+	}
+
+	// Use a prepuller implementation based on creating DaemonSets
+	// and block until all DaemonSets are ready; then we know for sure that all control plane images are cached locally
+	prepuller := upgrade.NewDaemonSetPrepuller(client, internalcfg)
+	upgrade.PrepullImagesInParallel(prepuller, flags.pullTimeout)
+
+	// Now; perform the upgrade procedure
+	if err := PerformUpgrade(flags, client, internalcfg); err != nil {
+		return err
+	}
+
+	// Upgrade RBAC rules and addons. Optionally, if needed, perform some extra task for a specific version
+	if err := upgrade.PerformPostUpgradeTasks(client, internalcfg, flags.newK8sVersion); err != nil {
+		return err
+	}
+
+	fmt.Println("")
+	fmt.Printf("[upgrade/successful] SUCCESS! Your cluster was upgraded to %q. Enjoy!\n", flags.newK8sVersionStr)
+	fmt.Println("")
+	fmt.Println("[upgrade/kubelet] Now that your control plane is upgraded, please proceed with upgrading your kubelets in turn.")
+
+	return nil
+}
+
+// ValidateFlags makes sure the arguments given to apply are valid
+func ValidateFlags(flags *applyFlags) error {
+	if len(flags.newK8sVersionStr) == 0 {
+		return fmt.Errorf("--version is a required parameter. Please specify the version you want to upgrade to")
```

Comment:
My personal preference, although I hope others weigh in on the expected UX here: the version to upgrade to is so central to the operation of upgrading (and is required) that I wouldn't even expect it to be a flag. I would actually expect this to just be a required argument, so you would do:

    $ kubeadm upgrade 1.8.0

instead of:

    $ kubeadm upgrade --version 1.8.0

To me, it's a little like having to type `ping --ip-address 127.0.0.1`.